### PR TITLE
[asl-carpenter] Introducing the ASLCarpenter testing tool

### PR DESCRIPTION
**Pull request by HadrienRenaud**
This PR introduces carpenter, an AST generator for ASL, working in two modes:
 - Enumerating all possible ASTs by increasing size
 - Randomly generating ASTs that respect some type-checking constraints

It can also work as a fuzzing tool, and as an execution tool for ASLRef.

Building and installing it is completely optional.

### CODE REVIEW

File: asllib/carpenter/README.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+dune build carpenter
+```
+
+## Usage
+
+See `carpenter --help` for a detailed view of options.
+
+Generating 10 tests:
+```
+dune exec carpenter -- generate random -o tmp 10
```
[the remark this answers is not in the document]
Comment (relokin) [reply asllib/carpenter/README.md:33]: Also this assumes that tmp exists.

File: asllib/carpenter/README.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+See `carpenter --help` for a detailed view of options.
+
+Generating 10 tests:
+```
+dune exec carpenter -- generate random -o tmp 10 
+```
+
+Executing them:
+```
+dune exec carpenter -- execute $(find tmp -name "$(date +"%Y-%m-%dT%H")*.asl")
```
Comment (relokin) [inline asllib/carpenter/README.md:38]: Are failures expected? I think they are but I wanted to check. Would it be worth to add a comment about it?

File: asllib/carpenter/README.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+```
+
+Fuzzing ASLRef on itself on 1000 tests:
+```
+dune exec carpenter -- fuzz enum -o tmp 1000
```
Comment (relokin) [inline asllib/carpenter/README.md:43]: Again it's a little unclear what I should expect in the output.

But overall maybe this is not meant to be a comprehensive guide so feel free to ignore.

File: asllib/carpenter/config.ml [hunk 4]
```diff
@@ -0,0 +1,538 @@
+module type Syntax = sig
```
Comment (relokin) [inline asllib/carpenter/config.ml:1]: Don't you need a copywrite header for this?

File: asllib/carpenter/ASTEnums.ml [hunk 5]
```diff
@@ -0,0 +1,427 @@
+(*
+ * SPDX-FileCopyrightText: Copyright 2022-2023 Arm Limited and/or its affiliates <open-source-office@arm.com>
```
Comment (relokin) [inline asllib/carpenter/ASTEnums.ml:2]: Did you make any changes in this file in 2024? Then you need to extend the copyright. This applies for all copyright headers.

```suggestion
 * SPDX-FileCopyrightText: Copyright 2022-2024 Arm Limited and/or its affiliates <open-source-office@arm.com>
```

File: asllib/carpenter/test/aslref.ml [hunk 6]
```diff
@@ -0,0 +1,56 @@
+let check_no_strange_error ast =
```
Comment (relokin) [inline asllib/carpenter/test/aslref.ml:1]: copyright header?

File: herdtools7.opam [hunk 7]
```diff
@@ -12,8 +12,8 @@ bug-reports: "http://github.com/herd/herdtools7/issues/"
 doc: "http://diy.inria.fr/doc/index.html"
 dev-repo: "git+https://github.com/herd/herdtools7.git"
 license: "CECILL-B"
-build: ["sh" "./dune-build.sh" "%{prefix}%"]
-install: ["sh" "./dune-install.sh" "%{prefix}%"]
+build: [make "build" "PREFIX=%{prefix}%"]
```
Comment (relokin) [inline herdtools7.opam:15]: Is this change needed because you introduce carpenter?

**Review by relokin — COMMENTED** [review]
LGTM, I have a few very superficial comments mostly on the README.md

File: asllib/carpenter/README.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+Dependencies:
+```bash
+opam install dune qcheck-core zarith re feat fix logs
+```
+
+And herdtools7 (see [herdtools7's readme](../../README.md)).
+
+Then:
+```bash
+dune build carpenter
```
Comment (relokin) [inline asllib/carpenter/README.md:24]: I just realised this need to be run from `asllib`, might be worth making it clear here.

File: asllib/carpenter/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (relokin) [reply asllib/carpenter/README.md:33]: `--progress` works now but some times this is getting stuck in the generation of a test (for example 8/10).

**Review by relokin — APPROVED** [review]
Thank you for all the changes!

This PR looks good to me but obviously I am not an expert. I am sure if anything comes up we can fix it later.

Feel free to merge.

File: asllib/carpenter/README.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+Both modes support constraints on syntax nodes available, and on execution or
+type-checking rules used by ASLRef to execute it.
+
+It can also serve as a fuzzing tool, or a mass execution tool.
+
+## Installation
+
+Dependencies:
+```bash
+opam install dune qcheck-core zarith re feat fix logs
```
Comment (maranget) [inline asllib/carpenter/README.md:17]: I followed the instructions and got this error:
```
File "asllib/carpenter/test/dune", line 5, characters 32-38:
5 |   (libraries carpenter_lib feat qcheck asllib))
                                    ^^^^^^
Error: Library "qcheck" not found.
```
I replaced `qcheck-core` by `qcheck` and ran `dune build asllib/carpenter` again and got this error:
```
File "asllib/carpenter/test/dune", line 5, characters 27-31:
5 |   (libraries carpenter_lib feat qcheck asllib))
                               ^^^^
Error: Library "feat" not found.
```
So I installed `feat`b resulting in the following `opam` command:
```
```suggestion
opam install dune qcheck zarith re feat fix logs feat
```

However compilation of carpenter still fails:
```
File "asllib/carpenter/dune", line 14, characters 35-43:
14 |  (libraries cmdliner carpenter_lib logs.cli))
                                        ^^^^^^^^
Error: Library "logs.cli" in /home/maranget/.opam/5.1.1/lib/logs is hidden
(unsatisfied 'exist_if').
```
Something may be wrong with my installation. In any case guidance would be appreciated.

```

Comment (maranget) [reply asllib/carpenter/README.md:17]: At some point cmdliner is also missing from the list of dependencies.

File: asllib/carpenter/main.ml [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+
+let pp_target fmt = function
+  | Itself -> Format.fprintf fmt "no"
+  | BInterp s -> Format.fprintf fmt "BInterp at %s" s
+
+let progress_frequency = 1
+
+let print_progress should_progress =
+  if should_progress then fun i n ->
+    if i mod progress_frequency == 0 then Printf.printf "\r%d/%d%!" i n else ()
```
Comment (maranget) [inline asllib/carpenter/main.ml:19]: Progress report output  will clobber valid output in case of redirection. Ideally, progress report could be disabled in case of redirection. However, I see no simple solution for now. As a mitigating measure, one can output the progress report on standard error:
```suggestion
    if i mod progress_frequency == 0 then Printf.eprintf "\r%d/%d%!" i n else ()
```

File: asllib/carpenter/main.ml [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+  else fun _i _n -> ()
+
+let hide_progress should_progress () =
+  if should_progress then
+    Printf.printf "\r                                          %!"
```
Comment (maranget) [inline asllib/carpenter/main.ml:24]: Again... (and add terminal linefeed):
```suggestion
    Printf.eprintf "\r                                          \r%!"
```

**Comment by maranget** [comment]
Build installations are now ok. I had a cosmetic remark. Appart from that, LGTM.

File: asllib/carpenter/main.ml (hunk 10 above)
Comment (maranget) [reply asllib/carpenter/main.ml:19]: Finally I found how to proceed:
```OCaml
let isatty chan =
   Unix.descr_of_out_channel chan |> Unix.isatty

let print_progress should_progress =
  if should_progress && isatty stderr then fun i n -> ...
```

File: asllib/carpenter/main.ml (hunk 11 above)
Comment (maranget) [reply asllib/carpenter/main.ml:24]: Aslo here:
```OCaml
let hide_progress should_progress () =
  if should_progress && isatty stderr then ...
```

File: asllib/carpenter/main.ml (hunk 10 above)
Comment (maranget) [reply asllib/carpenter/main.ml:19]: As for having progress output on stderr, users may want to redirect the normal output of the tool and still monitor progress. As for `isatty` you are right not to find this very useful, especially since de default behaviour is not outputting the progress reports. But well it does not harm at all for a small benefit...
